src/build/cargo.rs

```diff
@@ -61,7 +64,34 @@ pub(super) fn cargo(internals: &Internals) -> BuildResult {
     }
 }
 
-fn run_cargo(exec: RlsExecutor, rls_config: Arc<Mutex<Config>>, build_dir: PathBuf, out: Arc<Mutex<Vec<u8>>>) {
+fn run_cargo(compilation_cx: Arc<Mutex<CompilationContext>>,
+             rls_config: Arc<Mutex<Config>>,
+             vfs: Arc<Vfs>,
+             env_lock: Arc<EnvironmentLock>,
+             compiler_messages: Arc<Mutex<Vec<String>>>,
+             analyses: Arc<Mutex<Vec<Analysis>>>,
+             out: Arc<Mutex<Vec<u8>>>) {
+    // We need to fetch the outer and inner environment lock early here, because we need outer lock
+    // now to lock Environment for the scope of Cargo invocation and RlsExecutor needs to be
+    // initialized with the inner lock, since the RlsExecutor needs to hand it down during exec()
+    // callback, when calling linked rustc.
+    let outer_lock = env_lock.as_outer();
+    let (outer, inner) = outer_lock.get_lock();
+    // Lock early to guarantee synchronized access to read RUSTFLAGS env var
+    let env_lock_guard = outer.lock().unwrap();
```

Comment:
I imagine you would have your own guard struct that you would return, rather than returning either the mutex itself or a guard.